File: src/segment.ts
// Image segmentation — MaskFormer panoptic segmentation via @xenova/transformers
// Returns per-object masks with ADE20K class labels, mapped to 6 audio categories

import { pipeline, RawImage } from '@xenova/transformers';

type SegPipeline = Awaited<ReturnType<typeof pipeline<'image-segmentation'>>>;

let segPipeline: SegPipeline | null = null;
let modelFailed = false;

/**
 * Audio categories — each responds to different frequency bands.
 */
export const CATEGORY_COUNT = 6;

/**
 * ADE20K label name → audio category (0-5).
 *
 * 0 = BASS_SUBJECT (people, animals)
 * 1 = MID_ORGANIC (trees, plants, vegetation)
 * 2 = HIGH_SKY (sky, clouds, light sources)
 * 3 = BEAT_GROUND (ground, water, terrain)
 * 4 = MID_STRUCTURE (buildings, furniture, vehicles)
 * 5 = LOW_AMBIENT (walls, misc background)
 */
const LABEL_TO_CATEGORY: Record<string, number> = {};

// Category 0: BASS_SUBJECT — people, animals, living things
['person', 'animal', 'sculpture', 'statue', 'doll', 'dog', 'cat', 'horse',
 'bird', 'cow', 'sheep', 'elephant', 'bear', 'zebra', 'giraffe'].forEach(
  l => LABEL_TO_CATEGORY[l] = 0);

// Category 1: MID_ORGANIC — trees, plants, vegetation
['tree', 'grass', 'plant', 'field', 'palm', 'flower', 'bush', 'leaves',
 'vegetation', 'hedge', 'vine', 'moss', 'fern'].forEach(
  l => LABEL_TO_CATEGORY[l] = 1);

// Category 2: HIGH_SKY — sky, clouds, light
['sky', 'cloud', 'lamp', 'light', 'chandelier', 'sconce', 'sun', 'moon',
 'star', 'aurora', 'rainbow'].forEach(
  l => LABEL_TO_CATEGORY[l] = 2);

// Category 3: BEAT_GROUND — ground, water, terrain
['floor', 'road', 'sidewalk', 'earth', 'mountain', 'water', 'sea', 'river',
 'lake', 'path', 'sand', 'stairs', 'bridge', 'bench', 'dirt path', 'rock',
 'hill', 'waterfall', 'swimming pool', 'fountain', 'land', 'stairway',
 'runway', 'grandstand', 'pier'].forEach(
  l => LABEL_TO_CATEGORY[l] = 3);

// Category 4: MID_STRUCTURE — buildings, furniture, vehicles
['building', 'car', 'fence', 'door', 'table', 'chair', 'sofa', 'bed',
 'cabinet', 'desk', 'wardrobe', 'shelf', 'house', 'armchair', 'seat',
 'skyscraper', 'tower', 'bus', 'truck', 'van', 'ship', 'boat', 'airplane',
 'bicycle', 'minibike', 'signboard', 'column', 'railing', 'awning',
 'booth', 'canopy', 'tent', 'kitchen island', 'counter', 'countertop',
 'stove', 'oven', 'refrigerator', 'sink', 'bathtub', 'toilet',
 'television', 'computer', 'arcade machine', 'pool table', 'bar',
 'coffee table', 'fireplace', 'chest', 'bookcase', 'stage', 'conveyor belt',
 'washer', 'dishwasher', 'microwave', 'barrel', 'stool', 'ottoman',
 'swivel chair', 'hovel'].forEach(
  l => LABEL_TO_CATEGORY[l] = 4);

// Category 5: LOW_AMBIENT — walls, ceiling, misc (default for unlisted)
['wall', 'ceiling', 'windowpane', 'curtain', 'painting', 'mirror', 'rug',
 'cushion', 'base', 'box', 'pillow', 'screen door', 'blind', 'towel',
 'apparel', 'pole', 'bannister', 'escalator', 'bottle', 'buffet', 'poster',
 'case', 'cradle', 'blanket', 'hood', 'vase', 'tray', 'trash can', 'fan',
 'plate', 'monitor', 'crt screen', 'bulletin board', 'shower', 'radiator',
 'glass', 'clock', 'flag', 'bag', 'food', 'ball', 'step', 'tank',
 'trade name', 'pot', 'screen', 'book', 'plaything', 'basket'].forEach(
  l => LABEL_TO_CATEGORY[l] = 5);

function labelToCategory(label: string): number {
  // Direct lookup
  const lower = label.toLowerCase();
  if (lower in LABEL_TO_CATEGORY) return LABEL_TO_CATEGORY[lower];

  // Fuzzy matching for compound labels like "tree-merged"
  for (const [key, cat] of Object.entries(LABEL_TO_CATEGORY)) {
    if (lower.includes(key) || key.includes(lower)) return cat;
  }

  return 5; // default ambient
}

export interface SegmentResult {
  segments: Uint8Array;   // per-pixel audio category (0-5)
  count: number;          // always CATEGORY_COUNT (6)
  labels: string[];       // what was detected: "tree→cat1(25%)"
}

export async function estimateSegments(
  imageUrl: string,
  width: number,
  height: number,
  onStatus?: (msg: string) => void,
  depthMap?: Float32Array,
): Promise<SegmentResult> {
  if (!modelFailed) {
    try {
      if (!segPipeline) {
        onStatus?.('Loading segmentation model (~170MB)...');
        segPipeline = await pipeline(
          'image-segmentation',
          'onnx-community/maskformer-resnet50-ade20k-full',
        );
      }

      onStatus?.('Segmenting scene...');
      const results = await segPipeline(imageUrl);

      // Results: [{score, label, mask: RawImage}, ...]
      // Each mask is a binary RawImage at the input image resolution
      const items = Array.isArray(results) ? results : [results];

      // Build per-pixel category map
      // MaskFormer returns masks sorted by score (highest first)
      // We paint highest-score masks last so they win conflicts
      const segments = new Uint8Array(width * height);
      segments.fill(5); // default: ambient

      // Track what we found
      const labelCounts = new Map<string, number>();
      const totalPixels = width * height;

      // Sort by score ascending so highest-confidence masks paint last (win)
      const sorted = [...items].sort((a: any, b: any) => a.score - b.score);

      for (const item of sorted) {
        const { label, mask, score } = item as any;
        if (score < 0.3) continue; // skip low-confidence detections

        const cat = labelToCategory(label);
        const maskData = mask.data as Uint8Array;
        const maskW = mask.width;
        const maskH = mask.height;

        let pixelCount = 0;

        // Map mask pixels to output resolution
        for (let y = 0; y < height; y++) {
          for (let x = 0; x < width; x++) {
            const mx = Math.min(Math.floor((x / width) * maskW), maskW - 1);
            const my = Math.min(Math.floor((y / height) * maskH), maskH - 1);
            // mask.data is 1-channel, values are 0 or 255 (or boolean)
            if (maskData[my * maskW + mx] > 0) {
              segments[y * width + x] = cat;
              pixelCount++;
            }
          }
        }

        if (pixelCount > 0) {
          const existing = labelCounts.get(label) || 0;
          labelCounts.set(label, existing + pixelCount);
        }
      }

      // Build labels for display
      const labels: string[] = [];
      const sortedLabels = [...labelCounts.entries()].sort((a, b) => b[1] - a[1]);
      for (const [label, count] of sortedLabels) {
        if (count < 50) continue;
        const cat = labelToCategory(label);
        const pct = ((count / totalPixels) * 100).toFixed(0);
        labels.push(`${label}→cat${cat}(${pct}%)`);
      }

      console.log(`[segment] MaskFormer: ${items.length} objects, ${labelCounts.size} unique labels`);
      console.log(`[segment] Labels: ${labels.join(', ')}`);
      return { segments, count: CATEGORY_COUNT, labels };

    } catch (e) {
      const errMsg = (e as Error).message || String(e);
      console.error('[segment] MaskFormer failed:', errMsg);
      modelFailed = true;
      onStatus?.(`Seg failed: ${errMsg.slice(0, 80)}`);
      await new Promise(r => setTimeout(r, 2000));
    }
  }

  return depthFallback(width, height, depthMap);
}

function depthFallback(
  width: number,
  height: number,
  depthMap?: Float32Array,
): SegmentResult {
  const segments = new Uint8Array(width * height);

  if (depthMap && depthMap.length === width * height) {
    for (let i = 0; i < depthMap.length; i++) {
      const d = depthMap[i];
      if (d > 0.7) segments[i] = 0;
      else if (d > 0.5) segments[i] = 1;
      else if (d > 0.3) segments[i] = 4;
      else if (d > 0.15) segments[i] = 3;
      else segments[i] = 2;
    }
  } else {
    segments.fill(5);
  }

  return {
    segments,
    count: CATEGORY_COUNT,
    labels: [
      'close objects→cat0(depth)',
      'mid vegetation→cat1(depth)',
      'sky/far→cat2(depth)',
      'ground→cat3(depth)',
      'mid structures→cat4(depth)',
      'background→cat5(depth)',
    ],
  };
}

File: server/segment.py
"""Segmentation via SAM3 (Segment Anything with Concepts) — text-prompted, open vocabulary."""

import numpy as np
from PIL import Image
import logging
import torch

log = logging.getLogger(__name__)

# Text prompts for each audio category → SAM3 segments by concept
CATEGORY_PROMPTS = {
    0: ["person", "animal", "cat", "dog", "bird", "horse", "sculpture", "statue"],
    1: ["tree", "plant", "flower", "grass", "bush", "vegetation", "forest"],
    2: ["sky", "cloud", "sun", "moon", "star", "light", "lamp"],
    3: ["ground", "floor", "water", "sea", "river", "road", "sand", "snow", "rock", "mountain", "dirt"],
    4: ["building", "house", "car", "truck", "bridge", "fence", "chair", "table", "door", "window", "stairs"],
    5: ["wall", "curtain", "ceiling", "painting", "mirror", "screen"],
}

CAT_NAMES = ["BASS_SUBJECT", "MID_ORGANIC", "HIGH_SKY", "BEAT_GROUND", "MID_STRUCTURE", "LOW_AMBIENT"]

_sam3_model = None
_sam3_processor = None


def _load_sam3():
    """Load SAM3 model and processor."""
    global _sam3_model, _sam3_processor
    if _sam3_model is None:
        log.info("Loading SAM3...")
        from sam3.model_builder import build_sam3_image_model
        from sam3.model.sam3_image_processor import Sam3Processor

        _sam3_model = build_sam3_image_model()
        _sam3_processor = Sam3Processor(_sam3_model)
        log.info("SAM3 loaded.")
    return _sam3_model, _sam3_processor


def segment_image(
    image: Image.Image,
    dynamic_prompts: dict[int, list[str]] | None = None,
) -> tuple[np.ndarray, list[dict]]:
    """
    Run SAM3 text-prompted segmentation.
    If dynamic_prompts provided, use those instead of static CATEGORY_PROMPTS.
    Returns (segment_map uint8 HxW, detected_segments list of dicts).
    """
    model, processor = _load_sam3()
    w, h = image.size

    active_prompts = dynamic_prompts if dynamic_prompts else CATEGORY_PROMPTS
    log.info(f"Segmenting with {'dynamic' if dynamic_prompts else 'static'} prompts")

    # Default to ambient (category 5)
    segment_map = np.full((h, w), 5, dtype=np.uint8)
    detected = []

    # Set image once
    inference_state = processor.set_image(image)

    # Lower category numbers paint last (higher priority — subjects > ambient)
    for cat_id in sorted(active_prompts.keys(), reverse=True):
        prompts = active_prompts[cat_id]
        cat_pixel_count = 0

        for prompt_text in prompts:
            try:
                output = processor.set_text_prompt(
                    state=inference_state,
                    prompt=prompt_text,
                )
                masks = output.get("masks")
                scores = output.get("scores")

                if masks is None or len(masks) == 0:
                    continue

                # Process each detected instance
                for i, (mask, score) in enumerate(zip(masks, scores)):
                    if score < 0.3:
                        continue

                    # Convert mask to numpy bool array
                    if isinstance(mask, torch.Tensor):
                        mask_np = mask.cpu().numpy().astype(bool)
                    else:
                        mask_np = np.array(mask, dtype=bool)

                    # Handle different mask shapes
                    if mask_np.ndim == 3:
                        mask_np = mask_np[0]  # Take first channel
                    if mask_np.shape != (h, w):
                        mask_np = np.array(
                            Image.fromarray(mask_np.astype(np.uint8) * 255).resize((w, h), Image.NEAREST),
                            dtype=bool,
                        )

                    pixel_count = int(mask_np.sum())
                    if pixel_count < 100:  # Skip tiny detections
                        continue

                    segment_map[mask_np] = cat_id
                    cat_pixel_count += pixel_count

                    pct = pixel_count / (h * w) * 100
                    detected.append({
                        "label": prompt_text,
                        "category": cat_id,
                        "score": round(float(score), 3),
                        "coverage_pct": round(pct, 1),
                    })

            except Exception as e:
                log.warning(f"SAM3 prompt '{prompt_text}' failed: {e}")
                continue

        if cat_pixel_count > 0:
            total_pct = cat_pixel_count / (h * w) * 100
            log.info(f"Category {cat_id} ({CAT_NAMES[cat_id]}): {total_pct:.1f}% coverage")

    return segment_map, detected
